lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -5296,16 +5296,35 @@ FIRCircuitParser::parseModuleBody(DeferredModuleToParse &deferredModule) {
 
   // Convert print-encoded verifications after parsing.
 
-  // It is dangerous to modify IR in the walk, so accumulate printFOp to
-  // buffer.
-  SmallVector<PrintFOp> buffer;
-  deferredModule.moduleOp.walk(
-      [&buffer](PrintFOp printFOp) { buffer.push_back(printFOp); });
-
-  for (auto printFOp : buffer) {
-    auto result = circt::firrtl::foldWhenEncodedVerifOp(printFOp);
-    if (failed(result))
-      return result;
+  {
+    // Gather PrintFOps and then process to avoid walking while mutating.
+    SmallVector<PrintFOp> buffer;
+    deferredModule.moduleOp.walk(
+        [&buffer](PrintFOp printFOp) { buffer.push_back(printFOp); });
+
+    size_t numVerifPrintfs = 0;
+    std::optional<Location> printfLoc;
+    for (auto printFOp : buffer) {
+      auto loc = printFOp.getLoc();
+      auto result = circt::firrtl::foldWhenEncodedVerifOp(printFOp);
+      if (failed(result))
+        return failure();
+      if (*result) {
+        ++numVerifPrintfs;
+        if (!printfLoc)
+          printfLoc = loc;
+      }
+    }
+    if (numVerifPrintfs > 0) {
+      auto diag =
+          mlir::emitWarning(deferredModule.moduleOp.getLoc(),
+                            "module contains ")
+          << numVerifPrintfs
+          << " printf-encoded verification operation(s), which are deprecated "
+             "and will be removed in the future";
```

Comment:
Once per module might be annoying (but do we care since they are going away?) and could be once per circuit.

We might consider having a generic deprecated feature use collector.